### PR TITLE
Move trim functions (btrim, ltrim, rtrim) to datafusion_functions, make expr_fn API consistent

### LINKED ISSUE #9703: Move btrim, ltrim and rtrim to datafusion-functions
### Is your feature request related to a problem or challenge?

As part of https://github.com/apache/arrow-datafusion/issues/9285 move the trim functions (btrim, ltrim and rtrim) to the new datafusion-functions crate in the new structure. Update existing moved trim function to be btrim with an alias of 'trim' since they are the same impl.

### Describe the solution you'd like

Functions are migrated to the new crate and use the new patterns as described in https://github.com/apache/arrow-datafusion/issues/9286 & https://github.com/apache/arrow-datafusion/pull/9216

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by Omega359**
## Which issue does this PR close?

Closes #9703

## Rationale for this change

Function migration.
 
## What changes are included in this PR?

Code and tests.

## Are these changes tested?

Yes.

## Are there any user-facing changes?

Yes. The dataframe api now requires a vec![] parameter for all trim functions vs before where it seemed to be mixed between vec and not.

There was a mismatch between the sql interface, the dataframe api and the documentation around these functions. As you can see in the expr.slt file all the trim functions could be provided with a second argument which would be the string to use when trimming. However, the documentation and the dataframe api only seems to have that for btrim, not the other two functions.

This PR aligns the documentation and the dataframe with the sql so that all accept a second (optional) argument.

### CODE REVIEW

File: datafusion/functions/src/string/btrim.rs [hunk 1]
```diff
@@ -16,27 +16,27 @@
...
+
+/// Returns the longest string with leading and trailing characters removed. If the characters are not specified, whitespace is removed.
 /// btrim('xyxtrimyyx', 'xyz') = 'trim'
-pub fn btrim<T: OffsetSizeTrait>(args: &[ArrayRef]) -> Result<ArrayRef> {
+fn btrim<T: OffsetSizeTrait>(args: &[ArrayRef]) -> Result<ArrayRef> {
     general_trim::<T>(args, TrimType::Both)
 }
 
 #[derive(Debug)]
 pub(super) struct TrimFunc {
```
Comment (alamb) [inline datafusion/functions/src/string/btrim.rs:37]: It isn't a huge deal, but I found it a little confusing that the function is called `TrimFunc` but has a name of `"btrim"` and an alias of `"trim"`

I would expect either `BTrimFunc` or else the name of "trim" and an alias of `"btrim"`

File: datafusion/functions/src/string/mod.rs [hunk 2]
```diff
@@ -15,278 +15,63 @@
...
+//! "string" DataFusion functions
 
-/// Creates a function to identify the optimal return type of a string function given
-/// the type of its first argument.
-///
-/// If the input type is `LargeUtf8` or `LargeBinary` the return type is
-/// `$largeUtf8Type`,
-///
-/// If the input type is `Utf8` or `Binary` the return type is `$utf8Type`,
-macro_rules! get_optimal_return_type {
```
Comment (alamb) [inline datafusion/functions/src/string/mod.rs:39]: thank you for cleaning this up

File: datafusion/physical-expr/src/functions.rs [hunk 3]
```diff
@@ -752,70 +725,6 @@ mod tests {
             Int32Array
         );
         test_function!(BitLength, &[lit("")], Ok(Some(0)), i32, Int32, Int32Array);
-        test_function!(
```
Comment (alamb) [inline datafusion/physical-expr/src/functions.rs:755]: I double checked and  the functions are already tested in expr.slt (e.g. https://github.com/apache/arrow-datafusion/blob/47f4b5a67ac3b327764cbd4c0f42da7ac44854e5/datafusion/sqllogictest/test_files/expr.slt#L558-L574)

Thus removing these tests looks good to me 👍

File: docs/source/user-guide/sql/scalar_functions.md [hunk 4]
```diff
@@ -919,26 +922,25 @@ lpad(str, n[, padding_str])
 
 ### `ltrim`
 
-Removes leading spaces from a string.
+Trims the specified trim string from the beginning of a string.
```
Comment (alamb) [inline docs/source/user-guide/sql/scalar_functions.md:925]: Thank you for correcting and rationalizing the documentation

**Review by alamb — APPROVED** [review]
Thank you very much @Omega359  -- this looks great to me

File: datafusion/functions/src/string/btrim.rs (hunk 1 above)
Comment (Omega359) [reply datafusion/functions/src/string/btrim.rs:37]: Urgh, this was an oversight on my part when moving the Trim function here. I'll fix.

File: datafusion/physical-expr/src/functions.rs (hunk 3 above)
Comment (alamb) [reply datafusion/physical-expr/src/functions.rs:755]: ok. I actually think the sql coverage is the most important (as it ensures everything works end to end). That being said having additional unit test coverage is not a bad thing either

**Review by alamb — APPROVED** [review]
Thanks @Omega359

**Pull request merged**